sklearn/linear_model/cd_fast.pyx

```diff
@@ -161,127 +263,265 @@ def enet_coordinate_descent(np.ndarray[DOUBLE, ndim=1] w,
+                  <DOUBLE*>(X.data + i * n_samples * sizeof(DOUBLE)),
+                  1, <DOUBLE*>R.data, 1)
+
+        ### scaling tolerance
         # tol *= np.dot(y, y)
-        tol *= ddot(n_samples, <DOUBLE*>y.data, n_tasks,
-                    <DOUBLE*>y.data, n_tasks)
+        tol *= y_norm2
 
+        ### main loop
         for n_iter in range(max_iter):
+            ####################
+            # variable screening
+
+            do_gap = False
+            if (n_iter and w_max == 0.0  # heuristic termination criterion
+                    or d_w_max / w_max < d_w_tol):
+                # the biggest coordinate update of this iteration was smaller
+                # than the tolerance: check the duality gap as ultimate
+                # stopping criterion
+                do_gap = True
+
+            do_gap = (do_gap or (screening > 0 and n_iter % screening == 0)
+                      or (n_iter == max_iter - 1)) or (n_iter == 0)
+
+            if do_gap and screening > 0:  # Screening
+                gap = duality_gap(n_samples, n_features, n_tasks,
+                                  <DOUBLE*>X.data, <DOUBLE*>y.data, <DOUBLE*>R.data, <DOUBLE*>w.data,
+                                  <DOUBLE*>XtA.data, <DOUBLE*>dual_scaling.data,
+                                  alpha, beta, positive,
+                                  <np.int32_t*>disabled.data, n_features - n_active)
+
+                if (gap < tol) or (n_iter == max_iter - 1):
+                    # recompute non lazy gap as safeguard if screening is not correct
+                    # or if last iteration
+                    gap = duality_gap(n_samples, n_features, n_tasks,
+                                      <DOUBLE*>X.data, <DOUBLE*>y.data, <DOUBLE*>R.data, <DOUBLE*>w.data,
+                                      <DOUBLE*>XtA.data, <DOUBLE*>dual_scaling.data,
+                                      alpha, beta, positive,
+                                      <np.int32_t*>disabled.data, 0)
+                    # 0 because we do not trust the screening
+                    # XXX to do really?
+
+                    # break if we reached desired tolerance
+                    if gap < tol:
+                        break
+
+            if do_gap and screening == 0:
+                gap = duality_gap(n_samples, n_features, n_tasks,
+                                  <DOUBLE*>X.data, <DOUBLE*>y.data, <DOUBLE*>R.data, <DOUBLE*>w.data,
+                                  <DOUBLE*>XtA.data, <DOUBLE*>dual_scaling.data,
+                                  alpha, beta, positive,
+                                  <np.int32_t*>disabled.data, 0)
+
+                if gap < tol:
+                    # return if we reached desired tolerance
+                    break
+
+            if do_gap and screening > 0:  # Screening
+
+                r_screening = sqrt(2. * gap) / alpha
+                n_active = 0
+
+                for ii in range(n_features):  # Loop over coordinates
```

Comment:
the loop over features to screen some of them and only keep the indices of active ones in active_set